File: java/PassrodCripto.java
package com.solucionesnimrod.passrod.cripto;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.Base64;

/**
 * Núcleo criptográfico de PassRod v2 — implementación Java.
 *
 * Debe reproducir byte a byte los valores de {@code vectores/vectores.json}. Si
 * diverge de la implementación de TypeScript o Kotlin, una bóveda cifrada en un
 * cliente no se abrirá en otro; por eso los vectores se ejecutan como prueba.
 *
 * Sin dependencias externas: todo sale de la biblioteca estándar de Java.
 */
public final class PassrodCripto {

    public static final int VERSION_ESQUEMA = 2;
    public static final int ALG_AES_GCM = 0x01;
    public static final int PBKDF2_ITERACIONES = 600_000;

    private static final String PREFIJO_SALT = "passrod.v2|";
    private static final String PREFIJO_AAD = "passrod.v2|";
    private static final String INFO_ENC = "passrod.v2.enc";
    private static final String INFO_AUTH = "passrod.v2.auth";
    private static final String INFO_RECOVERY = "passrod.v2.recovery";

    private static final int NONCE_BYTES = 12;
    private static final int TAG_BITS = 128;

    private static final SecureRandom AZAR = new SecureRandom();

    private PassrodCripto() {}

    // ── Derivación ──────────────────────────────────────────────────────────

    /**
     * Salt determinista a partir del correo.
     *
     * Se deriva del correo y no de un aleatorio del servidor para que el cliente
     * pueda calcular la clave maestra ANTES de la primera petición.
     */
    public static byte[] saltDesdeEmail(String email) throws Exception {
        String normalizado = email.trim().toLowerCase();
        return MessageDigest.getInstance("SHA-256")
                .digest((PREFIJO_SALT + normalizado).getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Clave maestra con PBKDF2-SHA256.
     *
     * Se recibe {@code char[]} y no {@code String} a propósito: un String es
     * inmutable y queda en el heap hasta que pase el recolector, con lo que la
     * contraseña maestra puede acabar en un volcado de memoria o en el archivo
     * de intercambio. El array se puede sobrescribir al terminar.
     */
    public static byte[] derivarMK(char[] password, String email) throws Exception {
        byte[] salt = saltDesdeEmail(email);
        PBEKeySpec spec = new PBEKeySpec(password, salt, PBKDF2_ITERACIONES, 256);
        try {
            SecretKeyFactory f = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
            return f.generateSecret(spec).getEncoded();
        } finally {
            spec.clearPassword();
        }
    }

    /** HKDF-SHA256 con salt vacío (RFC 5869). */
    public static byte[] hkdf(byte[] ikm, String info, int largo) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");

        // extract
        mac.init(new SecretKeySpec(new byte[32], "HmacSHA256"));
        byte[] prk = mac.doFinal(ikm);

        // expand
        mac.init(new SecretKeySpec(prk, "HmacSHA256"));
        ByteArrayOutputStream salida = new ByteArrayOutputStream();
        byte[] bloque = new byte[0];
        byte[] infoBytes = info.getBytes(StandardCharsets.UTF_8);
        for (int contador = 1; salida.size() < largo; contador++) {
            mac.reset();
            mac.update(bloque);
            mac.update(infoBytes);
            mac.update((byte) contador);
            bloque = mac.doFinal();
            salida.write(bloque, 0, bloque.length);
        }
        return Arrays.copyOf(salida.toByteArray(), largo);
    }

    public static byte[] derivarSK(byte[] mk) throws Exception {
        return hkdf(mk, INFO_ENC, 32);
    }

    public static byte[] derivarAuthKey(byte[] mk) throws Exception {
        return hkdf(mk, INFO_AUTH, 32);
    }

    /** Lo ÚNICO que se envía al servidor para autenticarse. */
    public static String authHash(char[] password, String email) throws Exception {
        byte[] mk = derivarMK(password, email);
        byte[] ak = derivarAuthKey(mk);
        String salida = Base64.getEncoder().encodeToString(ak);
        limpiar(mk, ak);
        return salida;
    }

    // ── Formato de blob ─────────────────────────────────────────────────────

    /**
     * Datos autenticados asociados: atan el blob a su ubicación exacta.
     *
     * Sin esto, alguien con acceso de escritura a la base podría mover el
     * ciphertext de una credencial a otra fila y el descifrado seguiría
     * funcionando.
     */
    public static byte[] construirAAD(String tipo, Object idRecurso, Object idBoveda) {
        return (PREFIJO_AAD + tipo + "|" + idRecurso + "|" + idBoveda)
                .getBytes(StandardCharsets.UTF_8);
    }

    public static byte[] nuevoNonce() {
        byte[] n = new byte[NONCE_BYTES];
        AZAR.nextBytes(n);
        return n;
    }

    /** base64( VER(1) ‖ ALG(1) ‖ NONCE(12) ‖ CIPHERTEXT‖TAG(16) ) */
    public static String cifrar(byte[] clave, byte[] plano, byte[] aad, byte[] nonce)
            throws Exception {
        if (nonce.length != NONCE_BYTES) {
            throw new IllegalArgumentException("El nonce debe ser de 12 bytes");
        }
        Cipher c = Cipher.getInstance("AES/GCM/NoPadding");
        SecretKey k = new SecretKeySpec(clave, "AES");
        c.init(Cipher.ENCRYPT_MODE, k, new GCMParameterSpec(TAG_BITS, nonce));
        c.updateAAD(aad);
        byte[] ct = c.doFinal(plano);

        byte[] salida = new byte[2 + nonce.length + ct.length];
        salida[0] = (byte) VERSION_ESQUEMA;
        salida[1] = (byte) ALG_AES_GCM;
        System.arraycopy(nonce, 0, salida, 2, nonce.length);
        System.arraycopy(ct, 0, salida, 2 + nonce.length, ct.length);
        return Base64.getEncoder().encodeToString(salida);
    }

    public static byte[] descifrar(byte[] clave, String blob, byte[] aad) throws Exception {
        byte[] crudo = Base64.getDecoder().decode(blob);
        if (crudo.length < 2 + NONCE_BYTES + 16) {
            throw new IllegalArgumentException("Blob demasiado corto");
        }
        int version = crudo[0] & 0xFF;
        int alg = crudo[1] & 0xFF;
        // El byte de versión permite cambiar de algoritmo sin migrar toda la
        // base de golpe: cada blob dice cómo fue cifrado.
        if (version != VERSION_ESQUEMA) {
            throw new IllegalArgumentException("Versión de blob no soportada: " + version);
        }
        if (alg != ALG_AES_GCM) {
            throw new IllegalArgumentException("Algoritmo no soportado: " + alg);
        }

        byte[] nonce = Arrays.copyOfRange(crudo, 2, 2 + NONCE_BYTES);
        byte[] ct = Arrays.copyOfRange(crudo, 2 + NONCE_BYTES, crudo.length);

        Cipher c = Cipher.getInstance("AES/GCM/NoPadding");
        c.init(Cipher.DECRYPT_MODE, new SecretKeySpec(clave, "AES"),
               new GCMParameterSpec(TAG_BITS, nonce));
        c.updateAAD(aad);
        return c.doFinal(ct);
    }

    // ── Objetos de dominio ──────────────────────────────────────────────────

    public static String cifrarCredencial(byte[] vk, String json, Object idCredencial,
                                          Object idBoveda, byte[] nonce) throws Exception {
        return cifrar(vk, json.getBytes(StandardCharsets.UTF_8),
                      construirAAD("credencial", idCredencial, idBoveda), nonce);
    }

    public static String descifrarCredencial(byte[] vk, String blob, Object idCredencial,
                                             Object idBoveda) throws Exception {
        return new String(descifrar(vk, blob, construirAAD("credencial", idCredencial, idBoveda)),
                          StandardCharsets.UTF_8);
    }

    /** Clave de bóveda nueva. Aleatoria de verdad: nunca derivarla de nada. */
    public static byte[] nuevaClaveBoveda() {
        byte[] vk = new byte[32];
        AZAR.nextBytes(vk);
        return vk;
    }

    public static String envolverClaveBoveda(byte[] sk, byte[] vk, Object idBoveda, byte[] nonce)
            throws Exception {
        return cifrar(sk, vk, construirAAD("clave_boveda", idBoveda, idBoveda), nonce);
    }

    public static byte[] abrirClaveBoveda(byte[] sk, String envuelta, Object idBoveda)
            throws Exception {
        return descifrar(sk, envuelta, construirAAD("clave_boveda", idBoveda, idBoveda));
    }

    // ── Compartir bóvedas: RSA-2048-OAEP-SHA256 (§2.2 y §2.3) ───────────────
    //
    // Se eligió RSA y no X25519 porque WebCrypto, Java y Android lo traen de
    // serie; X25519 en WebCrypto es reciente y de disponibilidad desigual.
    //
    // El relleno de OAEP es aleatorio: dos envolturas de la misma clave dan
    // ciphertexts distintos. Por eso no se comparan byte a byte entre
    // implementaciones, sino descifrando.

    /** Nombre completo: sin él, Java usa SHA-1 en MGF1 y no interopera con WebCrypto. */
    private static final String RSA_OAEP = "RSA/ECB/OAEPWithSHA-256AndMGF1Padding";

    public static java.security.KeyPair generarParDeClaves() throws Exception {
        java.security.KeyPairGenerator g = java.security.KeyPairGenerator.getInstance("RSA");
        g.initialize(2048, AZAR);
        return g.generateKeyPair();
    }

    /**
     * Envuelve la clave de bóveda para otro usuario, con su clave pública.
     *
     * Es lo que permite compartir sin que el servidor participe: el dueño cifra
     * VK para el invitado y el servidor solo transporta el resultado.
     */
    public static String envolverParaUsuario(byte[] publicaSpki, byte[] vk) throws Exception {
        java.security.PublicKey pub = java.security.KeyFactory.getInstance("RSA")
                .generatePublic(new java.security.spec.X509EncodedKeySpec(publicaSpki));
        Cipher c = Cipher.getInstance(RSA_OAEP);
        c.init(Cipher.ENCRYPT_MODE, pub, parametrosOaep());
        return Base64.getEncoder().encodeToString(c.doFinal(vk));
    }

    /** Abre una clave de bóveda que envolvieron para mí. */
    public static byte[] abrirConPrivada(byte[] privadaPkcs8, String envuelta) throws Exception {
        java.security.PrivateKey priv = java.security.KeyFactory.getInstance("RSA")
                .generatePrivate(new java.security.spec.PKCS8EncodedKeySpec(privadaPkcs8));
        Cipher c = Cipher.getInstance(RSA_OAEP);
        c.init(Cipher.DECRYPT_MODE, priv, parametrosOaep());
        return c.doFinal(Base64.getDecoder().decode(envuelta));
    }

    /**
     * SHA-256 también en la función generadora de máscara.
     *
     * Java, con el nombre de algoritmo a secas, usa SHA-256 para el hash pero
     * deja SHA-1 en MGF1. WebCrypto usa SHA-256 en ambos, así que sin esto un
     * mensaje cifrado en el navegador no se abre en el escritorio.
     */
    private static javax.crypto.spec.OAEPParameterSpec parametrosOaep() {
        return new javax.crypto.spec.OAEPParameterSpec(
                "SHA-256", "MGF1",
                java.security.spec.MGF1ParameterSpec.SHA256,
                javax.crypto.spec.PSource.PSpecified.DEFAULT);
    }

    /** La privada nunca llega al servidor sin envolver. */
    public static String envolverClavePrivada(byte[] sk, byte[] privadaPkcs8, byte[] nonce)
            throws Exception {
        return cifrar(sk, privadaPkcs8, construirAAD("clave_privada", 0, 0), nonce);
    }

    public static byte[] abrirClavePrivada(byte[] sk, String envuelta) throws Exception {
        return descifrar(sk, envuelta, construirAAD("clave_privada", 0, 0));
    }

    // ── Recuperación ────────────────────────────────────────────────────────

    public static byte[] claveDeRecuperacion(String codigo) throws Exception {
        return hkdf(codigo.getBytes(StandardCharsets.UTF_8), INFO_RECOVERY, 32);
    }

    public static String envolverParaRecuperacion(String codigo, byte[] mk, byte[] nonce)
            throws Exception {
        byte[] rk = claveDeRecuperacion(codigo);
        String blob = cifrar(rk, mk, construirAAD("recovery", 0, 0), nonce);
        limpiar(rk);
        return blob;
    }

    public static byte[] recuperarMK(String codigo, String blob) throws Exception {
        byte[] rk = claveDeRecuperacion(codigo);
        byte[] mk = descifrar(rk, blob, construirAAD("recovery", 0, 0));
        limpiar(rk);
        return mk;
    }

    // ── Higiene ─────────────────────────────────────────────────────────────

    /** Sobrescribe material sensible en cuanto deja de hacer falta. */
    public static void limpiar(byte[]... buffers) {
        for (byte[] b : buffers) {
            if (b != null) Arrays.fill(b, (byte) 0);
        }
    }

    public static void limpiar(char[] password) {
        if (password != null) Arrays.fill(password, '\0');
    }
}

File: typescript/passrodCripto.ts
/**
 * Núcleo criptográfico de PassRod v2 — implementación para navegador y Node.
 *
 * Debe reproducir byte a byte los valores de `vectores/vectores.json`. Si esta
 * implementación diverge de la de Java o Kotlin, una bóveda cifrada en un
 * cliente no se abrirá en otro; por eso los vectores se ejecutan en CI.
 *
 * Todo el material sensible se deriva y se usa aquí, en el cliente. El servidor
 * solo recibe `auth_hash` y blobs opacos.
 */

let _cripto: Crypto | null = null;

/**
 * La WebCrypto del entorno, resuelta la PRIMERA VEZ QUE SE USA.
 *
 * Antes se resolvía al cargar el módulo, y con un `require('crypto').webcrypto`
 * de reserva para Node antiguo. Las dos cosas estorbaban:
 *
 *  - El `require` dentro de un módulo ESM lo intentan resolver los bundlers en
 *    tiempo de compilación aunque nunca se ejecute, y este fichero pasa a
 *    empaquetarse para la web y para la extensión de navegador.
 *  - Resolverla al cargar hace que el módulo REVIENTE AL IMPORTARSE si el
 *    entorno no tiene WebCrypto. Y este módulo se importa también en el
 *    servidor: la web corre en Amplify con Next en modo SSR, así que Node la
 *    carga aunque el cifrado ocurra siempre en el navegador. Un fallo ahí
 *    tumbaría el renderizado entero por una función que nadie iba a llamar.
 *
 * Resolviéndola de forma perezosa, el módulo se puede importar en cualquier
 * sitio y el error —si de verdad falta WebCrypto— aparece al cifrar, que es
 * cuando importa y donde se entiende.
 */
function obtenerCripto(): Crypto {
  if (_cripto) return _cripto;
  const c = globalThis.crypto;
  if (!c || !c.subtle) {
    throw new Error(
      'Este entorno no tiene WebCrypto (globalThis.crypto.subtle). PassRod ' +
        'necesita un navegador moderno o Node 18 o superior.'
    );
  }
  _cripto = c;
  return c;
}

// ── Parámetros del esquema (§2) ─────────────────────────────────────────────
export const VERSION_ESQUEMA = 2;
export const ALG_AES_GCM = 0x01;
export const PBKDF2_ITERACIONES = 600_000;

const PREFIJO_SALT = 'passrod.v2|';
const PREFIJO_AAD = 'passrod.v2|';
const INFO_ENC = 'passrod.v2.enc';
const INFO_AUTH = 'passrod.v2.auth';
const INFO_RECOVERY = 'passrod.v2.recovery';

const utf8 = new TextEncoder();

// ── Utilidades ──────────────────────────────────────────────────────────────

export function aBase64(datos: Uint8Array): string {
  let s = '';
  for (const b of datos) s += String.fromCharCode(b);
  return btoaSeguro(s);
}

export function deBase64(texto: string): Uint8Array {
  const s = atobSeguro(texto);
  const salida = new Uint8Array(s.length);
  for (let i = 0; i < s.length; i++) salida[i] = s.charCodeAt(i);
  return salida;
}

function btoaSeguro(s: string): string {
  return typeof btoa === 'function'
    ? btoa(s)
    : Buffer.from(s, 'binary').toString('base64');
}

function atobSeguro(s: string): string {
  return typeof atob === 'function'
    ? atob(s)
    : Buffer.from(s, 'base64').toString('binary');
}

/** Borra un buffer con material sensible. No es infalible en JS, pero acorta
 *  la ventana en que la clave sigue en memoria. */
export function limpiar(...buffers: Uint8Array[]): void {
  for (const b of buffers) b.fill(0);
}

// ── Derivación ──────────────────────────────────────────────────────────────

/**
 * Salt determinista a partir del correo.
 *
 * Se deriva del correo y no de un aleatorio del servidor para que el cliente
 * pueda calcular la clave maestra ANTES de la primera petición: así la
 * contraseña nunca necesita viajar ni esperar a nadie.
 */
export async function saltDesdeEmail(email: string): Promise<Uint8Array> {
  const normalizado = email.trim().toLowerCase();
  const h = await obtenerCripto().subtle.digest('SHA-256', utf8.encode(PREFIJO_SALT + normalizado));
  return new Uint8Array(h);
}

/**
 * Clave maestra con PBKDF2-SHA256.
 *
 * Es el algoritmo por defecto porque WebCrypto lo trae nativo, igual que Java y
 * Android. Argon2id resiste mejor el ataque por hardware dedicado, pero en el
 * navegador exige WASM; el campo `kdf_tipo` del usuario existe precisamente
 * para poder cambiarlo sin romper a quien ya está registrado.
 */
export async function derivarMK(password: string, email: string): Promise<Uint8Array> {
  const salt = await saltDesdeEmail(email);
  const base = await obtenerCripto().subtle.importKey(
    'raw', utf8.encode(password), { name: 'PBKDF2' }, false, ['deriveBits']);
  const bits = await obtenerCripto().subtle.deriveBits(
    { name: 'PBKDF2', salt, iterations: PBKDF2_ITERACIONES, hash: 'SHA-256' },
    base, 256);
  return new Uint8Array(bits);
}

/** HKDF-SHA256 con salt vacío (RFC 5869). */
export async function hkdf(ikm: Uint8Array, info: string, largo = 32): Promise<Uint8Array> {
  const base = await obtenerCripto().subtle.importKey('raw', ikm, { name: 'HKDF' }, false, ['deriveBits']);
  const bits = await obtenerCripto().subtle.deriveBits(
    { name: 'HKDF', hash: 'SHA-256', salt: new Uint8Array(0), info: utf8.encode(info) },
    base, largo * 8);
  return new Uint8Array(bits);
}

/** Clave de cifrado: nunca sale del cliente. */
export const derivarSK = (mk: Uint8Array) => hkdf(mk, INFO_ENC);

/** Clave de autenticación: su base64 es lo ÚNICO que se envía al servidor. */
export const derivarAuthKey = (mk: Uint8Array) => hkdf(mk, INFO_AUTH);

export async function authHash(password: string, email: string): Promise<string> {
  const mk = await derivarMK(password, email);
  const ak = await derivarAuthKey(mk);
  const salida = aBase64(ak);
  limpiar(mk, ak);
  return salida;
}

// ── Formato de blob ─────────────────────────────────────────────────────────

/**
 * Datos autenticados asociados: atan el blob a su ubicación exacta.
 *
 * Sin esto, alguien con acceso de escritura a la base podría mover el
 * ciphertext de una credencial a otra fila, o de una bóveda a otra, y el
 * descifrado seguiría funcionando.
 */
export function construirAAD(tipo: string, idRecurso: string | number,
                             idBoveda: string | number): Uint8Array {
  return utf8.encode(`${PREFIJO_AAD}${tipo}|${idRecurso}|${idBoveda}`);
}

/** base64( VER(1) ‖ ALG(1) ‖ NONCE(12) ‖ CIPHERTEXT‖TAG(16) ) */
export async function cifrar(clave: Uint8Array, plano: Uint8Array,
                             aad: Uint8Array, nonce?: Uint8Array): Promise<string> {
  const iv = nonce ?? obtenerCripto().getRandomValues(new Uint8Array(12));
  if (iv.length !== 12) throw new Error('El nonce debe ser de 12 bytes');

  const k = await obtenerCripto().subtle.importKey('raw', clave, { name: 'AES-GCM' }, false, ['encrypt']);
  const ct = new Uint8Array(
    await obtenerCripto().subtle.encrypt({ name: 'AES-GCM', iv, additionalData: aad, tagLength: 128 }, k, plano));

  const salida = new Uint8Array(2 + iv.length + ct.length);
  salida[0] = VERSION_ESQUEMA;
  salida[1] = ALG_AES_GCM;
  salida.set(iv, 2);
  salida.set(ct, 2 + iv.length);
  return aBase64(salida);
}

export async function descifrar(clave: Uint8Array, blob: string,
                                aad: Uint8Array): Promise<Uint8Array> {
  const crudo = deBase64(blob);
  if (crudo.length < 2 + 12 + 16) throw new Error('Blob demasiado corto');

  const version = crudo[0];
  const alg = crudo[1];
  // El byte de versión permite cambiar de algoritmo en el futuro sin migrar
  // toda la base de golpe: cada blob dice cómo fue cifrado.
  if (version !== VERSION_ESQUEMA) throw new Error(`Versión de blob no soportada: ${version}`);
  if (alg !== ALG_AES_GCM) throw new Error(`Algoritmo no soportado: ${alg}`);

  const iv = crudo.slice(2, 14);
  const ct = crudo.slice(14);
  const k = await obtenerCripto().subtle.importKey('raw', clave, { name: 'AES-GCM' }, false, ['decrypt']);
  const plano = await obtenerCripto().subtle.decrypt(
    { name: 'AES-GCM', iv, additionalData: aad, tagLength: 128 }, k, ct);
  return new Uint8Array(plano);
}

// ── Objetos de dominio ──────────────────────────────────────────────────────

export const cifrarCredencial = (vk: Uint8Array, credencial: unknown,
                                 idCredencial: string | number, idBoveda: string | number,
                                 nonce?: Uint8Array) =>
  cifrar(vk, utf8.encode(JSON.stringify(credencial)),
         construirAAD('credencial', idCredencial, idBoveda), nonce);

export async function descifrarCredencial<T>(vk: Uint8Array, blob: string,
                                             idCredencial: string | number,
                                             idBoveda: string | number): Promise<T> {
  const plano = await descifrar(vk, blob, construirAAD('credencial', idCredencial, idBoveda));
  return JSON.parse(new TextDecoder().decode(plano)) as T;
}

/** Clave de bóveda nueva. Aleatoria de verdad: nunca derivarla de nada. */
export const nuevaClaveBoveda = () => obtenerCripto().getRandomValues(new Uint8Array(32));

export const envolverClaveBoveda = (sk: Uint8Array, vk: Uint8Array,
                                    idBoveda: string | number, nonce?: Uint8Array) =>
  cifrar(sk, vk, construirAAD('clave_boveda', idBoveda, idBoveda), nonce);

export const abrirClaveBoveda = (sk: Uint8Array, envuelta: string,
                                 idBoveda: string | number) =>
  descifrar(sk, envuelta, construirAAD('clave_boveda', idBoveda, idBoveda));

// ── Compartir bóvedas: RSA-2048-OAEP-SHA256 (§2.2 y §2.3) ──────────────────
//
// Se eligió RSA y no X25519 porque WebCrypto, Java y Android lo traen de serie;
// X25519 en WebCrypto es reciente y de disponibilidad desigual.
//
// El relleno de OAEP es aleatorio: dos envolturas de la misma clave dan
// ciphertexts distintos. Es correcto, pero significa que no se pueden comparar
// byte a byte entre implementaciones — se comparan descifrando.

const RSA_OAEP = { name: 'RSA-OAEP', hash: 'SHA-256' } as const;

export interface ParDeClaves {
  publicaSpki: Uint8Array;   // va al servidor en claro: es pública
  privadaPkcs8: Uint8Array;  // va al servidor ENVUELTA con SK
}

/** Par de claves del usuario. Se genera una vez, en el registro. */
export async function generarParDeClaves(): Promise<ParDeClaves> {
  const par = await obtenerCripto().subtle.generateKey(
    { ...RSA_OAEP, modulusLength: 2048, publicExponent: new Uint8Array([1, 0, 1]) },
    true, ['encrypt', 'decrypt']) as CryptoKeyPair;
  return {
    publicaSpki: new Uint8Array(await obtenerCripto().subtle.exportKey('spki', par.publicKey)),
    privadaPkcs8: new Uint8Array(await obtenerCripto().subtle.exportKey('pkcs8', par.privateKey)),
  };
}

/**
 * Envuelve la clave de bóveda para otro usuario, usando su clave pública.
 *
 * Esto es lo que permite compartir sin que el servidor participe: el dueño
 * cifra VK para el invitado y el servidor solo transporta el resultado.
 */
export async function envolverParaUsuario(publicaSpki: Uint8Array,
                                          vk: Uint8Array): Promise<string> {
  const pub = await obtenerCripto().subtle.importKey('spki', publicaSpki, RSA_OAEP, false, ['encrypt']);
  return aBase64(new Uint8Array(await obtenerCripto().subtle.encrypt({ name: 'RSA-OAEP' }, pub, vk)));
}

/** Abre una clave de bóveda que envolvieron para mí. */
export async function abrirConPrivada(privadaPkcs8: Uint8Array,
                                      envuelta: string): Promise<Uint8Array> {
  const priv = await obtenerCripto().subtle.importKey('pkcs8', privadaPkcs8, RSA_OAEP, false, ['decrypt']);
  return new Uint8Array(
    await obtenerCripto().subtle.decrypt({ name: 'RSA-OAEP' }, priv, deBase64(envuelta)));
}

/** La privada nunca llega al servidor sin envolver. */
export const envolverClavePrivada = (sk: Uint8Array, privadaPkcs8: Uint8Array,
                                     nonce?: Uint8Array) =>
  cifrar(sk, privadaPkcs8, construirAAD('clave_privada', 0, 0), nonce);

export const abrirClavePrivada = (sk: Uint8Array, envuelta: string) =>
  descifrar(sk, envuelta, construirAAD('clave_privada', 0, 0));

// ── Recuperación ────────────────────────────────────────────────────────────

export const claveDeRecuperacion = (codigo: string) =>
  hkdf(utf8.encode(codigo), INFO_RECOVERY);

export async function envolverParaRecuperacion(codigo: string, mk: Uint8Array,
                                               nonce?: Uint8Array): Promise<string> {
  const rk = await claveDeRecuperacion(codigo);
  const blob = await cifrar(rk, mk, construirAAD('recovery', 0, 0), nonce);
  limpiar(rk);
  return blob;
}

export async function recuperarMK(codigo: string, blob: string): Promise<Uint8Array> {
  const rk = await claveDeRecuperacion(codigo);
  const mk = await descifrar(rk, blob, construirAAD('recovery', 0, 0));
  limpiar(rk);
  return mk;
}
